fairlearn/metrics/_metric_frame.py

```diff
@@ -0,0 +1,477 @@
+_FEATURE_LIST_NONSCALAR = "Feature lists must be of scalar types"
+_FEATURE_DF_COLUMN_BAD_NAME = "DataFrame column names must be strings. Name '{0}' is of type {1}"
+_DUPLICATE_FEATURE_NAME = "Detected duplicate feature name: '{0}'"
+_TOO_MANY_FEATURE_DIMS = "Feature array has too many dimensions"
+_SAMPLE_PARAM_KEYS_NOT_IN_FUNC_DICT = \
+    "Keys in 'sample_params' do not match those in 'metric_functions'"
+
+
+class MetricFrame:
+    """Contains a collection of group fairness metrics.
+
+    This class holds group fairness metrics calculated for any number of underlying
+    metrics. At least one sensitive feature must be supplied, which is used
+    to split the data into subgroups. The underlying metric(s) is(are) calculated
+    across the entire dataset (made available by the ``overall`` property) and
+    for each identified subgroup (made available by the ``by_group`` property).
+
+    The only limitations placed on the metric functions are that:
+
+    * The first two arguments they take must be ``y_true`` and ``y_pred`` arrays
+    * Any other arguments must be *sample* based, and will be split up along with
+      the ``y_true`` and ``y_pred`` arrays
+
+    The interpretation of the ``y_true`` and ``y_pred`` arrays is up to the
+    underlying metric - it is perfectly possible to pass in lists of class
+    probability tuples. The return value of the metric is also very flexible;
+    there is no problem around returning confusion matrices.
+
+    Once calculated, scalar group fairness metrics can be aggregated over
+    sensitive features, by taking differences or ratios,
+    and then reporting the maximum or minimum respectively.
+
+    This class also supports the concept of 'control features.' Like the sensitive
+    features, control features identify subgroups within the data, but
+    aggregations are not performed over the control features. Instead, the
+    aggregations produce a result for each subgroup identified by the control
+    feature(s). The name 'control features' refers to the statistical practice
+    of 'controlling' for a variable.
+
+    Parameters
+    ----------
+    metric_functions : callable or dict
+        The underlying metric functions which are to be calculated. This
+        can either be a single metric function or a dictionary of functions.
+        These functions must be callable as
+        ``fn(y_true, y_pred, **sample_params)``.
+        If there are any other arguments required (such as ``beta`` for
+        :func:`sklearn.metrics.fbeta_score`) then
+        :func:`functools.partial` must be used.
+
+    y_true : array_like
+        The true values
+
+    y_pred : array_like
+        The predicted values
+
+    sensitive_features : It's complicated
+        The sensitive features which should be used to create the subgroups.
+        At least one sensitive feature must be provided. There are several
+        possible input types. Acceptable 'unnamed' types are lists (of
```

Comment:
I think that "named" and "unnamed" distinction is not too relevant here. I think it would be clearer to instead separate out cases how to provide a single feature:
* `array1d-like`, which is one of: `list`, `array1d`, `pd.Series`

Or multiple features:
* `array2d`, `pd.DataFrame`, `dict` of array1d-like

I would talk about naming below that (unless there's a better place?).

Reply:
Redrafted.